### PR TITLE
DevTools: Show HTML tree

### PR DESCRIPTION
**Pull request by eerii**
This PR partly restores the HTML inspector functionality for the DevTools implementation. The inspector is probably the part of the devtools that has more [associated](https://searchfox.org/mozilla-central/source/devtools/server/actors/inspector) [actors](https://firefox-source-docs.mozilla.org/devtools/backend/actor-hierarchy.html). There were quite a few API changes since the last working implementation, so making it work again requires both fixing existing actors and creating new ones.

| **Existing actors** | **New actors** |
|---------------------|----------------|
| Inspector           | Accessibility  |
| Walker              | Access. Walker |
| Highligter          | Simulator      |
| Node                | CSS Properties |
| Page Style          | Reflow         |
| Layout              |                |
| Thread              |                |
| Network Parent      |                |

Some of the new actors and messages are placeholders that don't contain correct information yet, but they are required for the flow of messages to continue, since Firefox waits until a reply is issued for some of them to emit others.

image

Right now the DOM tree outline is showing, with two caveats that will be addressed in future PRs (since this has already gotten quite big):

- Node contents don't show up, for example, the text inside a paragraph. This is because this needs to be sent as a special inline content field in the node message which is not handled yet.
- Styles and CSS are not there yet since they require quite a few more work.

A few actors and message fields also seem to be outdated, namely Emulation, Performance and Timeline. Since Emulation was only a placeholder I removed it, but I kept the other two since they had work done in them that might be useful.

---
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes are a part of #32404
- [x] These changes do not require tests because there are no tests yet for DevTools

### CODE REVIEW

**Comment by nicoburns** [comment]
Amazing. As someone working on Servo's layout, I can't wait for this to land. At the moment if something isn't laying out correctly, it is not at all easy to work out why, and being able to interactively inspect the layout and the styles would be a complete game changer.

**Comment by mrobinson** [comment]
> Amazing. As someone working on Servo's layout, I can't wait for this to land. At the moment if something isn't laying out correctly, it is not at all easy to work out why, and being able to interactively inspect the layout and the styles would be a complete game changer.

In the meantime, you can also use `-Z dump-flow-tree` to when running Servo to see what the fragment tree looks like.

**Comment by nicoburns** [comment]
> In the meantime, you can also use `-Z dump-flow-tree` to when running Servo to see what the fragment tree looks like.

Oh, thanks. That looks really helpful. Although I'm still really looking forward to the interactive inspector, as I find that debug trees printed out in full are very verbose and hard to debug through (better than nothing though! And perhaps I can reduce my examples down to find the relevant parts of the tree)

**Comment by eerii** [comment]
Quick update on this, there is still work to be done but we finally have a preview of the DOM tree. It still doesn't show the contents or the style, but the node walking is now working.

image

Before going further I am going to focus on cleaning this pr, since it got a bit messy with all of the changes.

File: components/devtools/actors/inspector/walker.rs [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+                let _ = stream.write_json_packet(&msg);
+                ActorMessageStatus::Processed
+            },
+            _ => ActorMessageStatus::Ignored,
+        })
+    }
+}
+
+/// Recursively searchs for a child with the specified selector
+/// If it founds it, it returns a list with the child and all of its ascendents
```
Comment (mrobinson) [inline components/devtools/actors/inspector/walker.rs:215]:
```suggestion
/// Recursively searches for a child with the specified selector
/// If it is found, returns a list with the child and all of its ancestors.
```

File: components/devtools/actors/performance.rs [hunk 2]
```diff
@@ -2,6 +2,9 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
 
+// Is this actor still relevant?
```
Comment (mrobinson) [inline components/devtools/actors/performance.rs:5]:
```suggestion
// TODO: Is this actor still relevant?
```

File: components/devtools/actors/reflow.rs [hunk 3]
```diff
@@ -2,36 +2,47 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
 
+//! This actor is used for protocol purposes, it forwards the reflow events to clients
```
Comment (mrobinson) [inline components/devtools/actors/reflow.rs:5]:
```suggestion
//! This actor is used for protocol purposes, it forwards the reflow events to clients.
```

File: components/devtools/actors/reflow.rs [hunk 4]
```diff
@@ -2,36 +2,47 @@
...
 
-impl Actor for ProfilerActor {
+impl Actor for ReflowActor {
     fn name(&self) -> String {
         self.name.clone()
     }
 
+    /// The reflow actor can handle the following messages:
+    ///
+    /// - `start`: Does nothing yet. This doesn't need a reply like other messages
```
Comment (mrobinson) [inline components/devtools/actors/reflow.rs:25]:
```suggestion
    /// - `start`: Does nothing yet. This doesn't need a reply like other messages.
```

File: components/devtools/actors/reflow.rs [hunk 5]
```diff
@@ -2,36 +2,47 @@
...
-        _msg_type: &str,
+        msg_type: &str,
         _msg: &Map<String, Value>,
         _stream: &mut TcpStream,
         _id: StreamId,
     ) -> Result<ActorMessageStatus, ()> {
-        Ok(ActorMessageStatus::Ignored)
+        Ok(match msg_type {
+            "start" => {
+                // TODO: Create an observer on "reflows" events
```
Comment (mrobinson) [inline components/devtools/actors/reflow.rs:36]:
```suggestion
                // TODO: Create an observer on "reflows" events.
```

File: components/devtools/actors/timeline.rs [hunk 6]
```diff
@@ -2,6 +2,9 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
 
+// Is this actor still relevant?
```
Comment (mrobinson) [inline components/devtools/actors/timeline.rs:5]:
```suggestion
// TODO: Is this actor still relevant?
```

File: components/devtools/actors/watcher.rs [hunk 7]
```diff
@@ -142,6 +166,9 @@ impl Actor for WatcherActor {
     /// - `watchResources`: Start watching certain resource types. This sends
     /// `resource-available-form` events.
     ///
+    /// - `getNetworkParentActor`: Returns the network parent actor, it doesn't seem to do much at
+    /// the moment
```
Comment (mrobinson) [inline components/devtools/actors/watcher.rs:170]:
```suggestion
    /// - `getNetworkParentActor`: Returns the network parent actor. It doesn't seem to do much at
    /// the moment.
```

File: components/devtools/actors/watcher/network_parent.rs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+impl Actor for NetworkParentActor {
+    fn name(&self) -> String {
+        self.name.clone()
+    }
+
+    /// The network parent actor can handle the following messages:
+    ///
+    /// - `setSaveRequestAndResponseBodies`: Doesn't do anything yet
```
Comment (mrobinson) [inline components/devtools/actors/watcher/network_parent.rs:30]:
```suggestion
    /// - `setSaveRequestAndResponseBodies`: Doesn't do anything yet
```

File: components/devtools/actors/watcher/thread_configuration.rs [hunk 9]
```diff
@@ -0,0 +1,67 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+//! Liberally derived from <https://searchfox.org/mozilla-central/source/devtools/server/actors/thread-configuration.js>
+//! This actor manages the configuration flags that the devtools host can apply to the targets.
```
Comment (mrobinson) [inline components/devtools/actors/watcher/thread_configuration.rs:6]:
```suggestion
//! This actor manages the configuration flags that the devtools host can apply to threads.
```

**Review by mrobinson — APPROVED** [review]

**Comment by mrobinson** [comment]
I'm mainly rubber-stamping this one as there are many changes we need to make to the data structures and messages to become compatible with the newer version of the DevTools protocol. Nothing here seems out-of-line with that idea. Thanks!

**Pull request merged**